Add the Gumstix AeroCore2

smithandrewc opened the pull request.

LorenzMeier commented:
Nice, thanks!

LorenzMeier commented:
@smithandrewc Note the formatting fails that need fixing. For instructions see here:
https://travis-ci.org/PX4/Firmware/jobs/193487304

LorenzMeier commented on src/systemcmds/mtd/mtd.c:
```diff
@@ -303,6 +307,17 @@ mtd_start(char *partition_names[], unsigned n_partitions)
 
 		/* Create the partition */
 
+#ifdef CONFIG_ARCH_BOARD_AEROCORE2
```
This should be rebased on this PR and the partition table feature being used there:
https://github.com/PX4/Firmware/pull/6400

LorenzMeier commented:
Can you describe the issue you saw with Spektrum? I can completely believe that it had issues but would be good to understand the symptoms.

davids5 commented:
@LorenzMeier I would like to this PR to be rebased after the PR the removes all the #ifdefs https://github.com/PX4/Firmware/pull/6384 comes in.

smithandrewc commented:
The issue with the Spektrum was that it would only receive half the channels.  The channels would come in two frames.  The previous code would try and decode it but the missing channels would register as invalid so it would throw out all the data.  The PR changes the code so that it merges sequential frames to get all the channels.  This is done by using some global variables.

smithandrewc commented:
I've updated the code to incorporate #6400, #6384 and fix the formatting.  I adjusted the original commits.  I can create a new pull request once #6384 is merged.

LorenzMeier commented:
Needs a rebase on master. Note that we've moved board config related things to board_config.h

davids5 commented:
@smithandrewc - will you be able to rebase this on master soon?

smithandrewc commented:
Absolutely.  I had it prepared already, just need to do the last steps.  Once rebased, will I need to submit another pull request?

LorenzMeier commented:
No - force-pushing to your branch will update this pull request.

LorenzMeier commented:
@smithandrewc Please rebase.

smithandrewc commented:
@LorenzMeier Sorry I found an issue with mtd/dataman on the master branch and the AeroCore.  I'm trying to resolve it before I push.

dagar commented:
@smithandrewc did you get this working?

smithandrewc commented:
@dagar I got pulled away from the project for a while.  I plan to get back at it later this week.  Hopefully I'll have things fixed up shortly.

smithandrewc commented:
@dagar @LorenzMeier I finally got back to it and have force-pushed a new commit.  Thanks.

dagar commented:
Looks good. Can you review @davids5?

@smithandrewc if you add this to alt_firmware in the Makefile it will automatically build on semaphoreci. https://github.com/PX4/Firmware/blob/master/Makefile#L182

smithandrewc commented:
@dagar Sure ... should I add that to this commit and re-push?

dagar commented:
Yes, an additional commit is fine.

davids5 commented on src/drivers/boards/aerocore2/board_config.h:
```diff
@@ -0,0 +1,256 @@
+#define GPIO_TIM1_CH3IN		GPIO_TIM1_CH3IN_1
+#define DIRECT_INPUT_TIMER_CHANNELS  8
+
+/* spektrum satellite receiver input */
```
POWER_SPEKTRUM(_s) is not defined any longer.

Please See https://github.com/PX4/Firmware/blob/master/src/drivers/boards/px4fmu-v4/board_config.h#L265-L282 for an example of the new naming and usage.

davids5 commented on src/modules/dataman/dataman.c:
```diff
@@ -949,7 +949,15 @@ _file_initialize(unsigned max_offset)
 		close(dm_operations_data.file.fd);
 
 		if (incompat) {
+#ifdef CONFIG_ARCH_BOARD_AEROCORE2
```
@smithandrewc - As I understand it, this code ensures that the file format is compatible, with, the version that wrote.  Other platforms (px4fmuv2) use the FRAM as well for dataman storage, So I do not understand the comment or the need for the ifdef.  

Also as a rule we do not want want to have any ```CONFIG_ARCH_BOARD_``` in the system files. If this ifdef has a logical purpose (which is still TBD) we would want other boards to be able use the feature as BOARD_xxx. Like the naming that is done  https://github.com/PX4/Firmware/blob/master/src/drivers/boards/common/board_common.h or things like https://github.com/PX4/Firmware/blob/master/src/drivers/boards/tap-v1/board_config.h#L193 and https://github.com/PX4/Firmware/blob/master/src/drivers/boards/tap-v1/board_config.h#L245 

If you would help me to understand the usage I can suggest a name and where is should be placed.

smithandrewc replied:
This last one fails on a px4fmu-v2 as well.  I turned off autostart, at the console:
nsh>mtd start
nsh>mtd erase
Erasing /fs/mtd_params
Erased 8192 bytes
Erasing /fs/mtd_waypoints
Erased 8192 bytes
nsh>dataman start -f /fs/mtd_waypoints
INFO  [dataman] dataman file set to: /fs/mtd_waypoints
WARN  [dataman] Could not open data manager file /fs/mtd_waypoints
ERROR [dataman] dataman start failed

if you check the file system
nsh>ls /fs
/fs:
 bbr0
 bbr1
 bbr2
 mtd_params

It's missing mtd_waypoints.  This wasn't the case with the stable branch in January (not sure about now).  It was the case for the master branch in January, I didn't have time to investigate then.  For whatever reason, "unlink" is removing /fs/mtd_waypoints which is causing the "open" command to fail.

I'm investigating ...

davids5 replied:
@smithandrewc 

Ahh - yikes - this must have been broken in the recent work done on dataman!

@LorenzMeier - where is the preferred storage of waypoints now on the ramtron or the sdcard?

I think the issue is that the  ```/fs/mtd_waypoints``` is not a file it is a character device created on the block device.  In this case unlink is not the correct thing to do to a character device (driver).  The erase is done by writing 0xff to the device as in ```mtd_erase```

I imagine when the file name contains "mtd_" (or use stat() and look for buf->st_mode & S_IFCHR).)
we could call ```mtd_erase(k_data_manager_device_path, 1)``` instead of unlink.

smithandrewc replied:
By defining CONFIG_DISABLE_PSEUDOFS_OPERATIONS=y in the NuttX configs, it has fixed the issue.  This disables certain file operations that we probably don't want when using FRAM.

http://nuttx.org/Documentation/NuttXConfigVariables.html#CONFIG_DISABLE_PSEUDOFS_OPERATIONS

I've made the changes and re-pushed.

davids5 replied:
@smithandrewc - 

Great! I think that solution will work for this configuration. It will not on other platforms that support the hardfault_log.  

I have tested this with QGC.  (You will need the PID added to the inf file for windows,)

I did the Sensor calibration but I did not test the radio calibrations.

The Tone alarm is signaling error due to the lack of SDCARD and .
See:
```
# Run no SD alarm
if [ $LOG_FILE == /dev/null ]
```
in rcS.

Also, I noticed the LEDs are off. Did you want to add the LED lib and support the all 3 LEDs on the HW? If so, you will need the new led lib and an unfortunate #ifdef in commander.cpp control_status_leds(). See the ```cmake/configs/nuttx_px4fmu-v4_default.cmake``` for the led parts and ```src/drivers/boards/px4fmu-v4/px4fmu_led.c``` for a nice LED driver)

Thank you for getting this done.

smithandrewc replied:
@davids5 

Thanks for the tips on SDCARD and LED stuff.  I didn't notice the tone alarm for the SDCARD before.  I guess since it's always been there I always just got used to it.  That should be cleared now.  Also nice touch with the LEDs.  We weren't doing much with them so that's very helpful.  We only have 2 LEDs accessible to PX4 (other one connected to co-processor).  From what I can tell, the GREEN LED has more functionality so I've elected to drop the BLUE LED functionality and just go with GREEN.

Everything added and pushed.  Thanks.

davids5 commented on nuttx-configs/aerocore2/scripts/ld.script:
```diff
@@ -0,0 +1,159 @@
+
+MEMORY
+{
+    /* disabled due to silicon errata flash (rx)   : ORIGIN = 0x08004000, LENGTH = 2032K */
```
@smithandrewc 

Is the product being built with prior to rev 3 silicon, and subject to the [errata](https://pixhawk.org/help/errata)?

If not you can uses the full 2Mib and then enable all the modules that are applicable that come in px4fmu-v2 and px4fmu-v3 (no errata) and or px4fmu-v4 (no errata) builds.

Let us know if you plan to change this.  If not, then it is good to go as is.

smithandrewc replied:
New units are all v3 but older ones out in the wild do have older revisions.  I have one that's rev. Y.

We probably should keep it as is.  Thanks for pointing it out.
